packages/babel-parser/src/plugins/typescript/index.js

```diff
@@ -2281,6 +2283,16 @@ export default (superClass: Class<Parser>): Class<Parser> =>
       // "export declare" is equivalent to just "export".
       const isDeclare = this.eatContextual("declare");
 
+      if (
+        this.isContextual("declare") ||
+        !this.shouldParseExportDeclaration()
```

Comment:
nit: We should limit the extra check to `isDeclare === true`. When `isDeclare` is `false`, since no token was eaten, `shouldParseExportDeclaration()` is always `true` because

https://github.com/babel/babel/blob/446c70c6c3367565dafbcb98431442391eddd725/packages/babel-parser/src/parser/statement.js#L1799-L1804